Lighting

timoffex opened the pull request:
Added lighting.

timoffex committed:
MeshMaterial and lighting shader

The MeshMaterial class contains the information needed to properly
illuminate a point (for now, just the Phong model parameters).

The ‘onelight’ shader applies Phong lighting with one source of light.

timoffex committed:
Added normals & materials to RenderableObject

RenderableObject now has material and normal information for every
vertex.

timoffex committed:
Most beautiful GLSL program wrapper ever written

‘nuff said.

timoffex committed:
getPosition() for MeshViewCameras

Added a getPosition() method for cameras. This returns their
world-space position, which is necessary for lighting effects.

timoffex committed:
Default Mesh Materials

Added a “Soft” and “Metal” material, just for fun.

timoffex committed:
getPosition() amend

forgot to include this file

timoffex committed:
Fixed up lighting shaders

timoffex committed:
RenderableObject documentation

Added documentation to RenderableObject. Also fixed up
generateNormals() method.

Default material is MeshMaterial::Metal right now.

timoffex committed:
Changed MeshView to use lighting shaders

timoffex committed:
Modified loadOBJ() to add normals

loadOBJ() now uses normals information as well as position information.

The last section of loadOBJ() creates the pTriangleIndices array, which
indexes over UNIQUE PAIRS of position/normal indices. This is because
RenderableObject does not have separate index arrays for normals and
positions, which is because neither does OpenGL.

timoffex committed:
Removed unnecessary #defines

timoffex committed:
Added teapot

timoffex committed:
Merge branch 'master' into lighting

Fixed conflicts too.

Bdtrotte commented on WallsAndHoles/meshview.h.orig:
```diff
@@ -0,0 +1,94 @@
+#ifndef MESHVIEW_H
```
Stray file? Git rid of it!

Bdtrotte commented on WallsAndHoles/mainwindow.h:
```diff
@@ -34,6 +34,9 @@ private slots:
     void on_actionimport_triggered();
 
+protected:
+    MeshViewContainer *mMeshViewContainer;
```
Why is this here? Main window should remain empty. (MeshViewContainer is added as a dock widget by Editor)

Bdtrotte commented on WallsAndHoles/meshviewcontainer.cpp:
```diff
@@ -20,7 +20,7 @@ MeshViewContainer::MeshViewContainer(QWidget *parent) :
     //mMeshView->setScene(scene);
 
-    mMeshView->load(":/assets/cube.obj");
+    mMeshView->load(":/assets/teapot.obj");
```
Still a bad cube

timoffex replied:
We're not making Minecraft

Bdtrotte commented on WallsAndHoles/meshviewcontainer.cpp.orig:
```diff
@@ -0,0 +1,53 @@
+#include "meshviewcontainer.h"
```
Another stray file?

Bdtrotte reviewed (changes requested):
Some stray files and possibly lines. Good stuff otherwise!

timoffex committed:
Cleaned up MainWindow and got rid of .orig files

MainWindow still has the export/import actions. It may be worth moving
them elsewhere.
